### PR TITLE
feat(release): linux

### PR DESCRIPTION
**Pull request by daniellacosse**
As per [the second action item](https://docs.google.com/document/d/1TNBED78PnxUH987ZRXEZ0G2E-8vk92V6UbHLWS3Iq8Y/edit?resourcekey=0-oWsq__Qiyvp4RnZsAsTzww#heading=h.bhwmpv30h8yf) in the outline release user flow document, but only for the linux platform.

## Examples:

**Release Candidate Build**: https://github.com/daniellacosse/outline-client/actions/runs/1748056538

**Publish failure** (because tag was already published): https://github.com/daniellacosse/outline-client/runs/4944193101?check_suite_focus=true

**Publish success**: https://github.com/daniellacosse/outline-client/runs/4944324204?check_suite_focus=true
**`outline-releases` updated**: https://github.com/daniellacosse/outline-releases

### CODE REVIEW

File: .github/workflows/release_linux.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    - name: Install NPM Dependencies
+      run: npm ci
+
+    - name: Tag Release Commit
+      run: |
+        git config --global user.name "OutlineBot"
+        git config --global user.email "outlinebot@users.noreply.github.com"
+
+        git tag "linux-$(node scripts/get_version.mjs linux)"
```
Comment (daniellacosse) [inline .github/workflows/release_linux.yml:33]: This should yell if the tag already exists, preventing us from accidentally releasing the same version twice

File: .github/workflows/bulid_release_candidates.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+name: Build Release Candidates
+
+on:
+  push:
+    branches:
+      - master
+
+jobs:
+  linux:
```
Comment (daniellacosse) [inline .github/workflows/bulid_release_candidates.yml:9]: We'll add the jobs for the other platforms in this `jobs` list next.

File: .github/workflows/release_linux.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      uses: actions/checkout@v2.3.4
+      with:
+        repository: ${{ env.RELEASES_REPOSITORY }}
+        token: ${{ env.GITHUB_TOKEN }}
+        path: outline-releases
+    
+    - name: Push to Releases Repository
+      run: |
+        mv build/dist/Outline-Client.AppImage outline-releases/client/stable/Outline-Client.AppImage
+        mv build/dist/linux-beta.yml outline-releases/client/stable/linux-latest.yml
```
Comment (fortuna) [inline .github/workflows/release_linux.yml:54]: I think this is wrong. Why move beta to latest?
Also, prefer copy? Why move? I usually prefer to avoid mutability when possible.

File: .github/workflows/release_linux.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        RELEASE_VERSION="$(node scripts/get_version.mjs linux)""
+        cd outline-releases
+        git add client/stable/Outline-Client.AppImage client/stable/linux-latest.yml
+        git commit -m "release linux client ${RELEASE_VERSION}"
```
Comment (fortuna) [inline .github/workflows/release_linux.yml:58]: Capital letter
```suggestion
        git commit -m "Release linux client ${RELEASE_VERSION}"
```

File: .github/workflows/release_linux.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        git push origin master
```
Comment (fortuna) [inline .github/workflows/release_linux.yml:59]: I think we also tag the outline-releases

File: .github/workflows/release_linux.yml (hunk 4 above)
Comment (fortuna) [inline .github/workflows/release_linux.yml:58]: Can we make sure that the client is clean and there are no other edited non-committed file?
I want to avoid mistakes where we modify a file by mistake.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: .github/workflows/release_linux.yml (hunk 4 above)
Comment (daniellacosse) [reply .github/workflows/release_linux.yml:58]: All the commits in `outline-releases` are lower case - is that not desired?

Screen Shot 2022-01-12 at 11 12 18

Comment (daniellacosse) [reply .github/workflows/release_linux.yml:58]: Yep, I'll confirm this in my testing today.

File: .github/workflows/release_linux.yml (hunk 3 above)
Comment (daniellacosse) [reply .github/workflows/release_linux.yml:54]: HA, yeah I should be copying, here. Changed.

Hm, I should probably be uploading both linux-beta and linux-latest and copying the latter here.

File: .github/workflows/release_linux.yml (hunk 4 above)
Comment (daniellacosse) [reply .github/workflows/release_linux.yml:58]: (Sorry, testing this is next on my list right after confirming the android build fix)

File: .github/workflows/release_linux.yml [hunk 6]
```diff
@@ -52,9 +52,16 @@ jobs:
...
+        RELEASE_VERSION="$(node scripts/get_version.mjs linux)"
         cd outline-releases
         git add client/stable/Outline-Client.AppImage client/stable/linux-latest.yml
         git commit -m "Release linux client ${RELEASE_VERSION}"
         git tag "linux-${RELEASE_VERSION}"
-        git push origin master
\ No newline at end of file
+        git push origin master
+    
+    - name: Publish to Snap store
```
Comment (fortuna) [inline .github/workflows/release_linux.yml:62]: Is this working? Did you create a team account on the Snap store?
Let's move this to a separate PR.

File: src/electron/release_linux.action.sh [hunk 7]
```diff
@@ -55,7 +55,7 @@ npm run action src/electron/package_common linux --buildMode=release
 # provider since the S3 provider uses "virtual-hosted style" URLs (my-bucket.s3.amazonaws.com)
 # which can be blocked by DNS or SNI without taking down other buckets.
 electron-builder \
-  --linux \
+  --linux snap \
```
Comment (fortuna) [inline src/electron/release_linux.action.sh:58]: Revert for now

File: .github/workflows/build_and_test.yml [hunk 8]
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: Upload Linux Client
         uses: actions/upload-artifact@v2
         with:
-          name: unsigned-linux-client
+          name: debug-linux-client
```
Comment (fortuna) [inline .github/workflows/build_and_test.yml:44]: Perhaps we name it `client-[platform]-[mode]` so we can get better grouping in sorted order. (see all the stuff for the same platform together).

File: .github/workflows/build_and_test.yml [hunk 9]
```diff
@@ -68,7 +68,7 @@ jobs:
       - name: Upload Windows Client
         uses: actions/upload-artifact@v2
         with:
-          name: unsigned-windows-client
+          name: debug-windows-client
```
Comment (fortuna) [inline .github/workflows/build_and_test.yml:71]: I wonder if we should call this installer-windows-debug, since this is actually the installer of the client.

File: .github/workflows/release_linux.yml [hunk 10]
```diff
@@ -10,7 +10,9 @@ on:
 jobs:
   release: 
     env:
+      RELEASES_REPOSITORY: ${{ secrets.RELEASES_REPOSITORY }}
       GITHUB_TOKEN: ${{ secrets.ACCESS_TOKEN }}
+      STORE_LOGIN: ${{ secrets.SNAP_STORE_LOGIN }}
```
Comment (fortuna) [inline .github/workflows/release_linux.yml:15]: Revert

**Review by fortuna — CHANGES_REQUESTED** [review]

File: .github/workflows/build_release_candidates.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Install NPM Dependencies
+        run: npm ci
+
+      - name: Build Linux Release Candidate
+        run: npm run action src/electron/build_linux_release
+
+      - name: Upload Linux Release Candidate
+        uses: actions/upload-artifact@v2
+        with:
+          name: release-linux-${{ github.sha }}
```
Comment (daniellacosse) [inline .github/workflows/build_release_candidates.yml:37]: `github.sha` gives the full sha of the commit that triggered the build - we could truncate that, but there's always the slim chance the short sha refers to more than one commit...

File: .github/workflows/publish_linux_release_candidate.yml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        cp build/dist/Outline-Client.AppImage outline-releases/client/stable/Outline-Client.AppImage
+        cp build/dist/linux-latest.yml outline-releases/client/stable/linux-latest.yml
+        RELEASE_VERSION="$(node scripts/get_version.mjs linux)"
+        cd outline-releases
+        git add client/stable/Outline-Client.AppImage client/stable/linux-latest.yml
+        git commit -m "Release linux client ${RELEASE_VERSION}"
+        git tag "linux-${RELEASE_VERSION}"
+        git push origin master
+
+    # TODO(daniellacosse): S3 bucket?
```
Comment (daniellacosse) [inline .github/workflows/publish_linux_release_candidate.yml:62]: just want to make sure this S3 bucket is still a deploy target - https://github.com/Jigsaw-Code/outline-releases/blob/master/.github/workflows/upload_linux_client_to_s3.yml

File: .github/workflows/publish_linux_release_candidate.yml [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    
+    # TODO(daniellacosse): snap store
```
Comment (daniellacosse) [inline .github/workflows/publish_linux_release_candidate.yml:64]: snap _seems_ easy enough. we need to change the electron-builder platform option to `--linux snap`, then create a snap store account

File: src/electron/build_common.action.sh [hunk 14]
```diff
@@ -43,3 +43,5 @@ webpack \
 # TODO: make non-packaged builds work without this
 node scripts/environment_json.mjs "${PLATFORM}" \
     --buildMode="${BUILD_MODE}" > www/environment.json
+
+electron-icon-maker --input=resources/electron/icon.png --output=build
```
Comment (daniellacosse) [inline src/electron/build_common.action.sh:47]: so... this was the only difference between `build` and `package_common` - i didn't see the utility in having two different scripts, so i've merged them. lmk if there's something i'm ignorant of

File: .github/workflows/publish_linux_release_candidate.yml [hunk 15]
```diff
@@ -0,0 +1,70 @@
+name: Publish Linux Release Candidate
```
Comment (daniellacosse) [inline .github/workflows/publish_linux_release_candidate.yml:1]: I've chosen `publish` as the reserved word here instead of `release` as I don't want to confuse the build mode (`release`) with the action of publishing. 

Without this distinction, would "releasing" the binary mean building a release binary, or to actually push that binary to production?

This also required me to rename some scripts:

I've also renamed things so they're more consistent, we now have:
```
build, package_common => build_common
package_linux, package_windows => build_linux, build_windows
release_linux, release_windows => build_linux_release, build_windows_release
```

File: src/electron/build_linux_release.action.sh [hunk 16]
```diff
@@ -42,25 +42,18 @@ if ((STAGING_PERCENTAGE <= 0)) || ((STAGING_PERCENTAGE > 100)); then
...
+electron-builder --linux \
   --publish never \
   --config src/electron/electron-builder.json \
   --config.extraMetadata.version=$(node scripts/get_version.mjs linux) \
   --config.generateUpdatesFilesForAllChannels=true \
   --config.publish.provider=generic \
   --config.publish.url=https://s3.amazonaws.com/outline-releases/client/linux
 
-echo "stagingPercentage: $STAGING_PERCENTAGE" >> build/dist/$INFO_FILE_CHANNEL-linux.yml
+echo "stagingPercentage: $STAGING_PERCENTAGE" >> build/dist/beta-linux.yml
```
Comment (daniellacosse) [inline src/electron/build_linux_release.action.sh:58]: I don't really have any context on these staging_percentage things. do we need two separate files anymore?

Part of a larger convo - there's a large amount over overlap now between `build_linux` and `build_linux_release` actions. Main difference being this staging percentage thing and the additional build flags. How would you feel about me merging the two scripts so `build_linux --buildMode=release` effectively calls `build_linux_release`? I could also move these to node as well.

File: .github/workflows/release_linux.yml (hunk 6 above)
Comment (daniellacosse) [reply .github/workflows/release_linux.yml]: Ah, sorry, I just commented this out for later reference. Just re-requested ya.

File: .github/workflows/build_and_test.yml (hunk 9 above)
Comment (daniellacosse) [reply .github/workflows/build_and_test.yml:71]: client-installer? this is just the case for windows, right

File: .github/workflows/build_and_test.yml (hunk 8 above)
Comment (daniellacosse) [reply .github/workflows/build_and_test.yml:44]: yes, that'll play well down the line

File: src/electron/build_windows.action.sh [hunk 17]
```diff
@@ -38,8 +36,10 @@ if [[ -n ${SENTRY_DSN:-} ]]; then
   readonly SENTRY_URL="https://sentry.io/api/$PROJECT_ID/store/?sentry_version=7&sentry_key=$API_KEY"
 fi
 
+npm run action src/electron/build_common -- windows --buildMode="${BUILD_MODE}"
+
 # TODO: Move env.sh to build/electron/.
-cat >build/env.nsh <<EOF
+cat > build/env.nsh << EOF
```
Comment (fortuna) [inline src/electron/build_windows.action.sh:42]: I wonder how this is used and whether it's needed.

File: src/electron/build_common.action.sh (hunk 14 above)
Comment (fortuna) [reply src/electron/build_common.action.sh:41]: The semantics are confusing. Let's have a single build action (revert the name) that builds everything we need before electron-builder.

Move the package_windows logic to generate the env file here (and remove the existing duplication). It will have to be in an if-block based on the platform.

Once you call this script, all you need is to call electron-builder, which seems to take different options based on the mode.

File: src/electron/build_linux_release.action.sh (hunk 16 above)
Comment (fortuna) [reply src/electron/build_linux_release.action.sh:58]: One is for the beta channel and one for the stable. But I'm not sure whether we are using them.
For now, outputting them is fine. They will just be in sync.

We do want to use staged rollouts.

File: src/electron/build_linux.action.sh [hunk 18]
```diff
@@ -16,15 +16,19 @@
 BUILD_MODE=debug
```
Comment (fortuna) [inline src/electron/build_linux.action.sh:16]: Delete this file

File: src/electron/build_windows.action.sh [hunk 19]
```diff
@@ -28,8 +28,6 @@ for i in "$@"; do
     esac
```
Comment (fortuna) [inline src/electron/build_windows.action.sh:28]: Delete this file, as the logic will be in the build action.

File: src/electron/build_windows_release.action.sh [hunk 20]
```diff
@@ -43,16 +43,6 @@ if ((STAGING_PERCENTAGE <= 0)) || ((STAGING_PERCENTAGE > 100)); then
...
-else
-  INFO_FILE_CHANNEL="latest"
-fi
-
-npm run action src/electron/package_common windows \
-  --buildMode=release
-
 # Build the Sentry URL for the installer by parsing the API key and project ID from $SENTRY_DSN,
 # which has the following format: https://[32_CHAR_API_KEY]@sentry.io/[PROJECT_ID].
 readonly API_KEY=$(echo $SENTRY_DSN | awk -F/ '{print substr($3, 0, 32)}')
```
Comment (fortuna) [inline src/electron/build_windows_release.action.sh:48]: Delete these, as they will move to the build action.

File: src/electron/start.action.sh [hunk 21]
```diff
@@ -13,7 +13,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-npm run action src/electron/build -- "$@"
+npm run action src/electron/build_common -- "$@"
```
Comment (fortuna) [inline src/electron/start.action.sh:16]: Revert

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/electron/build_common.action.sh (hunk 14 above)
Comment (daniellacosse) [reply src/electron/build_common.action.sh:41]: Yep, that's basically what I was thinking. Exciting!

File: src/electron/build_linux_release.action.sh (hunk 16 above)
Comment (daniellacosse) [reply src/electron/build_linux_release.action.sh:58]: Of course - I'm going to update it to match the `--buildMode` flagging style as well

File: src/electron/build_windows.action.sh (hunk 17 above)
Comment (daniellacosse) [reply src/electron/build_windows.action.sh:42]: How do we confirm this?

File: src/electron/build.action.sh [hunk 22]
```diff
@@ -29,11 +34,41 @@ for i in "$@"; do
...
+!define SENTRY_URL "${SENTRY_URL:-}"
+EOF
+fi
+
+electron-builder $(node scripts/get_electron_build_flags.mjs ${PLATFORM} --buildMode=${BUILD_MODE})
+
+MANIFEST_POSTFIX=
+[[ "${PLATFORM}" == "linux" ]] && MANIFEST_POSTFIX="-linux"
+
+echo "stagingPercentage: $STAGING_PERCENTAGE" >> build/dist/beta${MANIFEST_POSTFIX}.yml
```
Comment (fortuna) [inline src/electron/build.action.sh:73]: Do we currently output this when we don't specify a staging percentage? I'm not sure what's the behavior if == 100. Perhaps we should just omit this?

File: scripts/get_electron_build_flags.mjs [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+  const platform = _[2];
+
+  console.log(
+    Object.entries(await getElectronBuildFlags(platform, buildMode)).map(
+      ([key, value]) => {
+        if (value === null) {
+          return `--${key}`;
+        }
+
+        return `--${key}=${value}`;
```
Comment (fortuna) [inline scripts/get_electron_build_flags.mjs:62]: This needs escaping. If a value has bad characters (e.g. `;`) it will break.

Perhaps use
```js
return `"${JSON.stringfy(`--\$${key}=\$${value}`)}"`
```
and put that in double quotes. Check if that's correct.

Comment (fortuna) [reply scripts/get_electron_build_flags.mjs:62]: It's probably easier to just have a list of flags though.
`args.push("--win")` or `args.push("'--config.win.certificateSubjectName=Jigsaw Operations LLC'")`, for example, and you certify that each push is valid shell.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: scripts/get_electron_build_flags.mjs (hunk 23 above)
Comment (daniellacosse) [reply scripts/get_electron_build_flags.mjs:62]: I did the list approach in the commit previous to this and can just revert it. I thought it would be nice to ensure there are no duplicate flags.

**Review by fortuna — APPROVED** [review]
Nice cleanup! Our build is becoming a lot more understandable!
